src/main.cpp

```diff
@@ -0,0 +1,333 @@
+        use_ilu = true;
+        break;
+
+      case 'a':
+        alpha = std::stod(optarg);
+        break;
+
+      case 'x':
+        tol = std::stod(optarg);
+        break;
+
+      case 'y':
+        tol_inner = std::stod(optarg);
+        break;
+
+      case '?':
+        pcout << err_msg << std::endl;
+        return 1;
+
+      default:
+        pcout << err_msg << std::endl;
+        return 1;
+    }
+  }
+
+  // Check if all required parameters are provided.
+  if (mesh_file_name.empty() || alpha <= 0 || alpha > 1) {
+    pcout << err_msg << std::endl;
+    return 1;
+  }
+
+  // Get the correct preconditioner and set solver options.
+  preconditioner_id preconditioner = SIMPLE;
+  switch (precondition_id) {
+    case 1:
+      preconditioner = BLOCK_DIAGONAL;
+      break;
+    case 2:
+      preconditioner = SIMPLE;
+      break;
+    case 3:
+      preconditioner = ASIMPLE;
+      break;
+    case 4:
+      preconditioner = YOSHIDA;
+      break;
+    default:
+      pcout << err_msg << std::endl;
+      return 1;
+  }
+  const SolverOptions solver_options(preconditioner, maxit, tol,
+                                     use_inner_solver, maxit_inner, tol_inner,
+                                     use_ilu, alpha);
+
+  // Run the chosen problem.
+  switch (problem_id) {
+    case 1: {
+      if (U_m == 0.0) U_m = 1.5;
+      Cylinder2D problem(mesh_file_name, degree_velocity, degree_pressure, T,
+                         deltat, U_m, varying_inlet, solver_options,
+                         compute_lift_drag);
+      problem.setup();
+      problem.solve();
+      break;
+    }
+
+    case 2: {
+      if (U_m == 0.0) U_m = 2.25;
+      Cylinder3D problem(mesh_file_name, degree_velocity, degree_pressure, T,
+                         deltat, U_m, varying_inlet, solver_options,
+                         compute_lift_drag);
+      problem.setup();
+      problem.solve();
+      break;
+    }
+
+    case 3: {
+      if (convergence_check) {
+        pcout << "Convergence check is being performed" << std::endl;
+        pcout << "===============================================" << std::endl;
+
+        constexpr unsigned int file_precision = 5;
+        constexpr unsigned int num_files = 4;
+
+        // We're setting T to 4*deltat so that only 4 time steps are performed,
+        // this is experimentally enough to get a precise enough result.
+        // For low values of T the stopping criterion on the residual can often
+        // lead to imprecise pressure solutions, especially for large meshes.
+        // Over multiple time steps, the phenomenon is mitigated.
+        const double T = 4.0 * deltat;
+        constexpr double nu = 0.01;
+
+        std::vector<double> h_values;
+        std::vector<double> errors_L2;
+        std::vector<double> errors_H1;
+
+        for (unsigned int i = 0; i < num_files; i++) {
+          // Extract the factor from the file name
+          size_t pos = mesh_file_name.find_last_of('-');
```

Comment:
You should always declare a variable as `const` if it is not meant to change:
```suggestion
          const size_t pos = mesh_file_name.find_last_of('-');
```